app/Services/InboundMail/InboundMailEngine.php

```diff
@@ -0,0 +1,279 @@
+            Log::info('E-Mail blocked, because the email was found on globalBlocklistEmails: ' . $from);
+            return true;
+        }
+
+        if (Cache::has('inboundMailBlockedSender:' . $from)) { // was marked as blocked before, so we block without any console output
+            return true;
+        }
+
+        // sender occured in more than 500 emails in the last 12 hours
+        $senderMailCountTotal = Cache::get('inboundMailCountSender:' . $from, 0);
+        if ($senderMailCountTotal >= 5000) {
+            Log::info('E-Mail blocked permanent, because the sender sended more than ' . $senderMailCountTotal . ' emails in the last 12 hours: ' . $from);
+            $this->blockSender($from);
+            $this->saveMeta($from, $to);
+            return true;
+        }
+        if ($senderMailCountTotal >= 1000) {
+            Log::info('E-Mail blocked, because the sender sended more than ' . $senderMailCountTotal . ' emails in the last 12 hours: ' . $from);
+            $this->saveMeta($from, $to);
+            return true;
+        }
+
+        // sender sended more than 50 emails to the wrong mailbox in the last 6 hours
+        $senderMailCountUnknownRecipent = Cache::get('inboundMailCountSenderUnknownRecipent:' . $from, 0);
+        if ($senderMailCountUnknownRecipent >= 50) {
+            Log::info('E-Mail blocked, because the sender sended more than ' . $senderMailCountUnknownRecipent . ' emails to the wrong mailbox in the last 6 hours: ' . $from);
+            $this->saveMeta($from, $to);
+            return true;
+        }
+
+        // wrong recipent occurs in more than 100 emails in the last 12 hours, so the processing is blocked
+        $mailCountUnknownRecipent = Cache::get('inboundMailCountUnknownRecipent:' . $to, 0); // @turbo124 maybe use many to save resources in case of spam with multiple to addresses each time
+        if ($mailCountUnknownRecipent >= 200) {
+            Log::info('E-Mail blocked, because anyone sended more than ' . $mailCountUnknownRecipent . ' emails to the wrong mailbox in the last 12 hours. Current sender was blocked as well: ' . $from);
+            $this->blockSender($from);
+            $this->saveMeta($from, $to);
+            return true;
+        }
+
+        return false;
+    }
+    public function blockSender(string $from)
+    {
+        Cache::add('inboundMailBlockedSender:' . $from, true, now()->addHours(12));
+
+        // TODO: ignore, when known sender (for heavy email-usage mostly on isHosted())
+        // TODO: handle external blocking
+    }
+    public function saveMeta(string $from, string $to, bool $isUnknownRecipent = false)
+    {
+        // save cache
+        Cache::add('inboundMailCountSender:' . $from, 0, now()->addHours(12));
+        Cache::increment('inboundMailCountSender:' . $from);
+
+        if ($isUnknownRecipent) {
+            Cache::add('inboundMailCountSenderUnknownRecipent:' . $from, 0, now()->addHours(6));
```

Comment:
For all cache ref's, you will want to prefix with $company->company_key otherwise you may have collisions when processing with other companies?

Reply:
if we change the cache keys to this, we dont have a global blocking mechanism anymore. Do you want this? Or do we have to change this only for a few entries, like unknown recipient?

Reply:
@paulwer did we make progress on this one? we'd need to have a way to filter based on company for this action.

Reply:
as mentioed, i have concerns, moving this scoped to a company, because it would be easier for spamming. Maybe we have to do a double implementation.
f.ex. add an additional check for unknownSenders of a company scope

Reply:
@paulwer becomes tricky for hosted implementation, as you could block a person in one company, and it blocks in every company.

Reply:
yes you are right, so i think we have to do both.


global blocking with larger contingents until someone gets blocked to prevent denial of service when to many requests on purpose or by bug are arriving.

company level blocking to reduce spam or inapropiate behavior on recieving and creating expenses for a single company with smaller contingents until someone gets blocked.

can you may send me your ideas, which we should have in place for both types of blocking and i will check or may implement additional mechanisms?